cabal-install/Distribution/Client/IndexUtils/Timestamp.hs

```diff
@@ -0,0 +1,117 @@
+-- |
+-- Module      :  Distribution.Client.IndexUtils.Timestamp
+-- Copyright   :  (c) Herbert Valerio Riedel
+-- License     :  BSD3
+--
+-- Timestamp type used in package indexes
+
+module Distribution.Client.IndexUtils.Timestamp
+    ( Timestamp
+    , nullTimestamp
+    , epochTimeToTimestamp
+    , timestampToUTCTime
+    , utcTimeToTimestamp
+    ) where
+
+import qualified Codec.Archive.Tar.Entry    as Tar
+import           Control.DeepSeq
+import           Data.Char                  (isDigit)
+import           Data.Int                   (Int64)
+import           Data.Time                  (UTCTime (..), fromGregorianValid,
+                                             makeTimeOfDayValid, showGregorian,
+                                             timeOfDayToTime, timeToTimeOfDay)
+import           Data.Time.Clock.POSIX      (posixSecondsToUTCTime,
+                                             utcTimeToPOSIXSeconds)
+import           Distribution.Compat.Binary
+import qualified Distribution.Compat.ReadP  as ReadP
+import           Distribution.Text
+import qualified Text.PrettyPrint           as Disp
+
+-- | UNIX timestamp (expressed in seconds since unix epoch, i.e. 1970).
+newtype Timestamp = TS Int64 -- Tar.EpochTime
+                  deriving (Eq,Ord,Enum,NFData)
+
+epochTimeToTimestamp :: Tar.EpochTime -> Timestamp
+epochTimeToTimestamp = TS
+
+timestampToUTCTime :: Timestamp -> Maybe UTCTime
+timestampToUTCTime (TS t)
+  | t == minBound  = Nothing
+  | otherwise      = Just $ posixSecondsToUTCTime (fromIntegral t)
+
+utcTimeToTimestamp :: UTCTime -> Maybe Timestamp
+utcTimeToTimestamp utct
+  | minTime <= t, t <= maxTime  = Just (TS (fromIntegral t))
+  | otherwise                   = Nothing
+  where
+    maxTime = toInteger (maxBound :: Int64)
+    minTime = toInteger (succ minBound :: Int64)
+    t :: Integer
+    t = round . utcTimeToPOSIXSeconds $ utct
+
+-- | Pretty-prints 'Timestamp' in ISO8601/RFC3339 format
+-- (e.g. @"2017-12-31T23:59:59Z"@)
+--
+-- Returns empty string for 'nullTimestamp'
+showTimestamp :: Timestamp -> String
+showTimestamp ts = case timestampToUTCTime ts of
+    Nothing          -> ""
+    Just UTCTime{..} -> showGregorian utctDay ++ ('T':showTOD utctDayTime) ++ "Z"
+  where
+    showTOD = show . timeToTimeOfDay
+
+instance Binary Timestamp where
+    put (TS t) = put t
+    get = TS `fmap` get
+
+instance Text Timestamp where
+    disp = Disp.text . showTimestamp
+
+    parse = parsePosix ReadP.+++ parseUTC
+      where
+        -- | Parses unix timestamps, e.g. @"\@1474626019"@
+        parsePosix = do
+            _ <- ReadP.char '@'
+            t <- parse -- FIXME
+            return (TS $ fromIntegral (t :: Int))
+
+        -- | Parses ISO8601/RFC3339-style UTC timestamps,
+        -- e.g. @"2017-12-31T23:59:59Z"@
+        --
+        -- TODO: support numeric tz offsets; allow to leave off seconds
+        parseUTC = do
```

Comment:
And similarly to not using `formatTime`, you've probably got good reasons not to use `readPTime`, but it's worth explaining in a comment.